### DISCUSSION TITLE
Incorect temprerature with BME688

### DISCUSSION THREAD

**Discussion by florian-terisse** [mattjlewis/diozero#201] [Q&A]
Hi all,

I try to get the temperature with diosero framework from a BME688 senson (adafruit) connected to a PI5.

The temperature releved is incorrect (34°)

When i run the bosh python test script (BME68X_Environmental_Sensor_code/RaspberryPi/Python/examples/read-all.py) the temperature seems correct (22,98).

If i initialise the BME688 with the diorero framework during python script is runing, temperature releved by the python script change :

Logs of the python script :
22.98 C,1004.79 hPa,43.78 %RH,156527.05594619384 Ohms
22.98 C,1004.79 hPa,43.78 %RH,156670.74663402693 Ohms
22.98 C,1004.79 hPa,43.78 %RH,156383.6285888821 Ohms
34.81 C,683.08 hPa,100.00 %RH                             <-- At this instant, i run the java program with diozero framework
34.81 C,683.08 hPa,100.00 %RH
34.81 C,683.08 hPa,100.00 %RH
34.81 C,683.08 hPa,100.00 %RH
34.81 C,683.08 hPa,100.00 %RH

I'm sure the problem is on my usage of the diozero framework but i can't find where i'm wrong ...

Someone have an idea ?

Thank your for your help and sorry for my bat english..

**Comment by EAGrahamJr** [comment]
I also note that these values from the library code:

```
Data [newData=true, heaterTempStable=false, gasMeasurementValid=false, gasMeasurementIndex=0, measureIndex=255, temperature=33.600376, pressureHPa=675.69244, humidity=100.0, gasResistance=0.0, idacHeatRegVal=0, heaterResistance=0, gasWaitMs=0]
```

And they do not change over time.

I think there's a trick to initializing the heater and the sampling, but I'm at a loss to figure it out myself...

**Comment by EAGrahamJr** [comment]
Well, I feel a _little_ like an idiot -- @florian-terisse take a look in the [`diozero-sampleapps`](https://github.com/mattjlewis/diozero/tree/main/diozero-sampleapps) subproject: there's a `BME68xTest` that properly sets up the sensor and outputs the corrected data.

```
java -cp diozero-sampleapps-1.4.1.jar com.diozero.sampleapps.BME68xTest 1 119
```
This is assuming you have the same one I do (I2C/STEMMA connection) and your system is similar to:
```
$ /sbin/i2cdetect -y 1                                                                                                                                                                          1 ↵
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- 3c -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- 77
```
(The 3c is an OLED display)

**Comment by mattjlewis** [comment]
I thought I had all this tested - was just digging out a couple of BME688 devices I have spare but not connected. Let me know if I need to initialise with more sensible default values.

**Reply by EAGrahamJr** [reply]
It certainly doesn't look like it -- the problem was knowing **how** to use the darn thing, since it's not really "plug and play".

```
Reading [1]: Idx: 136. Temperature: 30.14 C. Pressure: 1,008.20 hPa. Relative Humidity: 32.17 %rH. Gas Idx: 6. Gas Resistance: 156,814.70 Ohms. IDAC: 9.25 mA. Gas Wait: 1 (ms or multiplier). (heater stable: true, gas valid: true).
```
Noting that Adafruit's code applies a "temperature correction", this is right on with other sensors, as well as in the same range as output by the Adafruit CircuitPython code.

**Comment by EAGrahamJr** [comment]
Just one question: where did you get the Air Quality calculation from? This is starting from https://github.com/mattjlewis/diozero/blob/129bcba12154ac63a38e90d96d45f0e55a892eee/diozero-sampleapps/src/main/java/com/diozero/sampleapps/BME68xTest.java#L240

**Reply by mattjlewis** [reply]
https://github.com/pimoroni/bme680-python/blob/main/examples/indoor-air-quality.py

**Comment by EAGrahamJr** [comment]
#207 should be gtg -- I've been running it since the PR and the data has been independently verified against the sample code above (over time).
